

<!DOCTYPE html>


<html lang="en" data-theme="light">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Tutorials &#8212; BulkDGD  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "light";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Tutorials';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="The core subpackage" href="core.html" />
    <link rel="prev" title="The command-line interface" href="Command-line%20interface.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="light">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  <div class="navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    <p class="title logo__title">BulkDGD  documentation</p>
  
</a></div>
    
  </div>
  
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Getting%20started.html">
                        Getting started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Installation.html">
                        Installing bulkDGD
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Command-line%20interface.html">
                        The command-line interface
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Tutorials
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="core.html">
                        The core subpackage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="utils.html">
                        The utils subpackage
                      </a>
                    </li>
                
                </div>
            </div>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
        </div>
      
      
        <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Getting%20started.html">
                        Getting started
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Installation.html">
                        Installing bulkDGD
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="Command-line%20interface.html">
                        The command-line interface
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        Tutorials
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="core.html">
                        The core subpackage
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="utils.html">
                        The utils subpackage
                      </a>
                    </li>
                
                </div>
            </div>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumbs">
  <ul class="bd-breadcrumbs" role="navigation" aria-label="Breadcrumb">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">Tutorials</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tutorials">
<h1>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading">#</a></h1>
<p>In this section, you will find examples of how to use the <code class="docutils literal notranslate"><span class="pre">bulkDGD</span></code> package for different tasks.</p>
<p>If you want an overview of the command-line utilities available in <code class="docutils literal notranslate"><span class="pre">bulkDGD</span></code>, see the <a class="reference internal" href="Command-line%20interface.html"><span class="doc">Command-line interface</span></a> section.</p>
<p>Here, we will showcase the usage of the package’s functions in the context of larger analysis scripts.</p>
<p>Specifically, we provide detailed tutorials to:</p>
<ul class="simple">
<li><p>Find the best representations in latent space for a new set of samples (<a class="reference internal" href="#tutorial-1-finding-the-best-representations-for-a-set-of-new-samples"><span class="xref myst">Tutorial 1</span></a>).</p></li>
<li><p>Perform differential expression analysis between a set of “treated” samples (for instance, cancer samples) and their corresponding “untreated” samples (“normal” samples) found using the DGD model (<a class="reference internal" href="#tutorial-2-differential-expression-analysis"><span class="xref myst">Tutorial 2</span></a>).</p></li>
<li><p>Perform principal component analysis (PCA) on a set of representations and plot the results (<a class="reference internal" href="#tutorial-3-principal-component-analysis"><span class="xref myst">Tutorial 3</span></a>).</p></li>
</ul>
<p>The data needed to reproduce the tutorials can be found in the <code class="docutils literal notranslate"><span class="pre">bulkDGD/tutorials/data</span></code> directory, while the scripts containing the code necessary to run the tutorials and the output files can be found in each tutorial’s directory.</p>
<section id="tutorial-1-finding-the-best-representations-for-a-set-of-new-samples">
<h2>Tutorial 1 - Finding the best representations for a set of new samples<a class="headerlink" href="#tutorial-1-finding-the-best-representations-for-a-set-of-new-samples" title="Permalink to this heading">#</a></h2>
<p>All the code necessary to run this tutorial can be found in the <code class="docutils literal notranslate"><span class="pre">tutorial_1.py</span></code> script.</p>
<section id="the-input-samples">
<h3>The input samples<a class="headerlink" href="#the-input-samples" title="Permalink to this heading">#</a></h3>
<p>In this tutorial, we are going to find the best representations in the latent space defined by the DGD model for the samples in the <code class="docutils literal notranslate"><span class="pre">samples.csv</span></code> file. The file has the following structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">,</span><span class="n">ENSG00000187634</span><span class="p">,</span><span class="n">ENSG00000188976</span><span class="p">,</span><span class="n">ENSG00000187961</span><span class="p">,</span><span class="n">ENSG00000187583</span><span class="p">,</span><span class="o">...</span><span class="p">,</span><span class="n">tissue</span>
<span class="mi">1627</span><span class="p">,</span><span class="mi">80736</span><span class="p">,</span><span class="mi">275265</span><span class="p">,</span><span class="mi">52208</span><span class="p">,</span><span class="mi">2088</span><span class="p">,</span><span class="o">...</span><span class="p">,</span><span class="n">testis</span>
<span class="mi">111</span><span class="p">,</span><span class="mi">44899</span><span class="p">,</span><span class="mi">176358</span><span class="p">,</span><span class="mi">65177</span><span class="p">,</span><span class="mi">2660</span><span class="p">,</span><span class="o">...</span><span class="p">,</span><span class="n">adipose_visceral_omentum</span>
<span class="mi">555</span><span class="p">,</span><span class="mi">60662</span><span class="p">,</span><span class="mi">381897</span><span class="p">,</span><span class="mi">90671</span><span class="p">,</span><span class="mi">24486</span><span class="p">,</span><span class="o">...</span><span class="p">,</span><span class="n">breast_mammary_tissue</span>
<span class="o">...</span>
</pre></div>
</div>
<p>As we can see, each row contains the expression data for a specific sample. The first column contains the samples’ unique names, IDs, or indexes, while the rest of the columns contain either the expression data for a specific gene (identified by its Ensembl ID) or additional information about the samples. In our case, the last column identifies the tissue from which the sample comes.</p>
<p>Before finding the representations for these samples, we want to make sure that the genes whose expression data are reported correspond to the genes used to train the DGD model and that these genes are reported in the correct order in the file. We can do it by using the <code class="docutils literal notranslate"><span class="pre">preprocess_samples</span></code> in the <code class="docutils literal notranslate"><span class="pre">bulkDGD.utils.dgd</span></code> module.</p>
<p>First, we set the logging so that every message above and including the <code class="docutils literal notranslate"><span class="pre">INFO</span></code> level gets reported, to have a better idea of what the program is doing.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the logging module</span>
<span class="kn">import</span> <span class="nn">logging</span> <span class="k">as</span> <span class="nn">log</span>

<span class="c1"># Set the logging options</span>
<span class="n">log</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span> <span class="o">=</span> <span class="s2">&quot;INFO&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then, we load our CSV file as a data frame.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import Pandas and the &#39;bulkDGD.utils.dgd&#39; module</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">bulkDGD.utils.dgd</span> <span class="k">as</span> <span class="nn">dgd</span>

<span class="c1"># Load the samples into a data frame</span>
<span class="n">df_samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="c1"># Name of/path to the file</span>
                         <span class="s2">&quot;samples.csv&quot;</span><span class="p">,</span>
                         <span class="c1"># Column separator used in the file</span>
                         <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">,</span>
                         <span class="c1"># Name or numerical index of the column</span>
                         <span class="c1"># containing the samples&#39; names/IDs/indexes)</span>
                         <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
                         <span class="c1"># Name or numerical index of the row</span>
                         <span class="c1"># containing the columns&#39; names</span>
                         <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Then, we can preprocess the samples.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Preprocess the samples</span>
<span class="n">df_preproc</span><span class="p">,</span> <span class="n">genes_excluded</span><span class="p">,</span> <span class="n">genes_missing</span> <span class="o">=</span> \
    <span class="n">dgd</span><span class="o">.</span><span class="n">preprocess_samples</span><span class="p">(</span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">df_samples</span><span class="p">)</span>

<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="mi">1</span> <span class="n">column</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">containing</span> <span class="n">additional</span> <span class="n">information</span> <span class="p">(</span><span class="ow">not</span> <span class="n">gene</span> <span class="n">expression</span> <span class="n">data</span><span class="p">)</span> <span class="n">were</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">data</span> <span class="n">frame</span><span class="p">:</span> <span class="n">tissue</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Now</span> <span class="n">looking</span> <span class="k">for</span> <span class="n">duplicated</span> <span class="n">samples</span><span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">No</span> <span class="n">duplicated</span> <span class="n">samples</span> <span class="n">were</span> <span class="n">found</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Now</span> <span class="n">looking</span> <span class="k">for</span> <span class="n">missing</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">columns</span> <span class="n">containing</span> <span class="n">gene</span> <span class="n">expression</span> <span class="n">data</span><span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">No</span> <span class="n">missing</span> <span class="n">values</span> <span class="n">were</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">columns</span> <span class="n">containing</span> <span class="n">gene</span> <span class="n">expression</span> <span class="n">data</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Now</span> <span class="n">looking</span> <span class="k">for</span> <span class="n">duplicated</span> <span class="n">genes</span><span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">No</span> <span class="n">duplicated</span> <span class="n">genes</span> <span class="n">were</span> <span class="n">found</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">In</span> <span class="n">the</span> <span class="n">data</span> <span class="n">frame</span> <span class="n">containing</span> <span class="n">the</span> <span class="n">pre</span><span class="o">-</span><span class="n">processed</span> <span class="n">samples</span><span class="p">,</span> <span class="n">the</span> <span class="n">columns</span> <span class="n">containing</span> <span class="n">gene</span> <span class="n">expression</span> <span class="n">data</span> <span class="n">will</span> <span class="n">be</span> <span class="n">ordered</span> <span class="n">according</span> <span class="n">to</span> <span class="n">the</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">genes</span> <span class="n">used</span> <span class="n">to</span> <span class="n">train</span> <span class="n">the</span> <span class="n">DGD</span> <span class="n">model</span> <span class="p">(</span><span class="n">which</span> <span class="n">can</span> <span class="n">be</span> <span class="n">found</span> <span class="ow">in</span> <span class="s1">&#39;/Users/testuser/programs/bulkDGD/data/model/training_genes.txt&#39;</span><span class="p">)</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">In</span> <span class="n">the</span> <span class="n">data</span> <span class="n">frame</span> <span class="n">containing</span> <span class="n">the</span> <span class="n">pre</span><span class="o">-</span><span class="n">processed</span> <span class="n">samples</span><span class="p">,</span> <span class="n">the</span> <span class="n">columns</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">data</span> <span class="n">frame</span> <span class="n">which</span> <span class="n">did</span> <span class="ow">not</span> <span class="n">contain</span> <span class="n">gene</span> <span class="n">expression</span> <span class="n">data</span><span class="p">,</span> <span class="k">if</span> <span class="nb">any</span> <span class="n">were</span> <span class="n">present</span><span class="p">,</span> <span class="n">will</span> <span class="n">appended</span> <span class="k">as</span> <span class="n">the</span> <span class="n">last</span> <span class="n">columns</span> <span class="n">of</span> <span class="n">the</span> <span class="n">data</span> <span class="n">frame</span><span class="p">,</span> <span class="ow">and</span> <span class="n">appear</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">same</span> <span class="n">order</span> <span class="k">as</span> <span class="n">they</span> <span class="n">did</span> <span class="ow">in</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">data</span> <span class="n">frame</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">All</span> <span class="n">genes</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">samples</span> <span class="n">are</span> <span class="n">part</span> <span class="n">of</span> <span class="n">the</span> <span class="nb">set</span> <span class="n">of</span> <span class="n">genes</span> <span class="n">used</span> <span class="n">to</span> <span class="n">train</span> <span class="n">the</span> <span class="n">DGD</span> <span class="n">model</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">All</span> <span class="n">genes</span> <span class="n">used</span> <span class="n">to</span> <span class="n">train</span> <span class="n">the</span> <span class="n">DGD</span> <span class="n">model</span> <span class="n">were</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">samples</span><span class="o">.</span>
</pre></div>
</div>
<p>By inspecting the log messages, we can see that the function has looked for duplicated samples, duplicated genes, and missing values in the columns containing gene expression data. In fact, duplicated samples, duplicated genes, and missing values must be dealt with before preprocessing the samples.</p>
<p>Then, the function informs us that the columns containing gene expression data were re-ordered according to the list of genes used to train the DGD model, and that all the columns containing additional information about the samples (in our case, the <code class="docutils literal notranslate"><span class="pre">tissue</span></code> column) will be the last columns of the output data frame.</p>
<p>Finally, the function checks that all genes in the input samples are among those used to train the DGD model, and that all genes used in the DGD model are found in the input samples.</p>
<p>The function returned three objects:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">df_preproc</span></code> is a data frame containing the preprocessed samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genes_excluded</span></code> is a list containing the Ensembl IDs of the genes that were found in the input samples but are not part of the set of genes used to train the DGD model. These genes are not present in <code class="docutils literal notranslate"><span class="pre">df_preproc</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genes_missing</span></code> is a list containing the Ensembl IDs of the genes that are part of the set of genes used to train the DGD model but were not found in the input samples. These genes are added to <code class="docutils literal notranslate"><span class="pre">df_preproc</span></code> with a count of 0 for all samples.</p></li>
</ul>
</section>
<section id="the-configuration-files">
<h3>The configuration files<a class="headerlink" href="#the-configuration-files" title="Permalink to this heading">#</a></h3>
<p>We also need a configuration file containing the specifications of the DGD model and another configuration file with the options to fine-tune the search for the best representations in latent space.</p>
<p>For this, we will use the <code class="docutils literal notranslate"><span class="pre">bulkDGD/configs/model/config_model.yaml</span></code> and the <code class="docutils literal notranslate"><span class="pre">bulkDGD/configs/representations/config_rep.yaml</span></code> files, respectively.</p>
<p>We assume the files have been copied to the current working directory for the tutorial.</p>
<p>We can load the configurations using the <code class="docutils literal notranslate"><span class="pre">utils.misc.get_config_model</span></code> and the <code class="docutils literal notranslate"><span class="pre">utils.misc.get_config_rep</span></code> functions. Therefore, we load the <code class="docutils literal notranslate"><span class="pre">misc</span></code> module first.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the &#39;bulkDGD.utils.misc&#39; module</span>
<span class="kn">from</span> <span class="nn">bulkDGD.utils</span> <span class="kn">import</span> <span class="n">misc</span>
</pre></div>
</div>
<section id="the-model-s-configuration">
<h4>The model’s configuration<a class="headerlink" href="#the-model-s-configuration" title="Permalink to this heading">#</a></h4>
<p>We can load the model’s configuration using the <code class="docutils literal notranslate"><span class="pre">utils.misc.get_config_model</span></code> function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the model&#39;s configuration</span>
<span class="n">config_model</span> <span class="o">=</span> <span class="n">misc</span><span class="o">.</span><span class="n">get_config_model</span><span class="p">(</span><span class="s2">&quot;config_model.yaml&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Once loaded, each configuration consists of a dictionary of options.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># We can inspect the configuration</span>
<span class="n">config_model</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">{</span><span class="s2">&quot;gmm&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;pth_file&quot;</span><span class="p">:</span> <span class="s2">&quot;/Users/testuser/programs/bulkDGD/data/model/gmm.pth&quot;</span><span class="p">,</span> <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;dim&#39;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span> <span class="s2">&quot;n_comp&quot;</span><span class="p">:</span> <span class="mi">45</span><span class="p">,</span> <span class="s2">&quot;cm_type&quot;</span><span class="p">:</span> <span class="s2">&quot;diagonal&quot;</span><span class="p">,</span> <span class="s2">&quot;log_var_params&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">},</span> <span class="s2">&quot;mean_prior&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;softball&quot;</span><span class="p">,</span> <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;radius&quot;</span><span class="p">:</span> <span class="mi">7</span><span class="p">,</span> <span class="s2">&quot;sharpness&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">}}},</span> <span class="s2">&quot;dec&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;pth_file&quot;</span><span class="p">:</span> <span class="s2">&quot;/Users/testuser/programs/tests/dgd_tests/dec.pth&quot;</span><span class="p">,</span> <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;n_neurons_input&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span> <span class="s2">&quot;n_neurons_hidden&quot;</span><span class="p">:</span> <span class="mi">500</span><span class="p">,</span> <span class="s2">&quot;n_neurons_hidden2&quot;</span><span class="p">:</span> <span class="mi">8000</span><span class="p">,</span> <span class="s2">&quot;n_neurons_output&quot;</span><span class="p">:</span> <span class="mi">16883</span><span class="p">,</span> <span class="s2">&quot;r_init&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}}}</span>
</pre></div>
</div>
<p>Here, we provide a brief description of every option. You can find similar descriptions in the comments inside the configuration file.</p>
<p><code class="docutils literal notranslate"><span class="pre">&quot;gmm&quot;</span></code> maps to a dictionary containing the options to get the trained Gaussian mixture model:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;pth_file&quot;</span></code> defines the PyTorch file containing the pre-trained Gaussian mixture model. If a file path is specified, the corresponding file will be used. If <code class="docutils literal notranslate"><span class="pre">&quot;default&quot;</span></code> is specified in the configuration file, the <code class="docutils literal notranslate"><span class="pre">bulkDGD/data/model/gmm.pth</span></code> file will be used, and the absolute path to this file will appear in the configuration.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;options&quot;</span></code> maps to the options needed to initialize the Gaussian mixture model:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;dim&quot;</span></code> indicates the dimensionality of the Gaussian mixture model.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_comp&quot;</span></code> is the number of components in the Gaussian mixture.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;cm_type&quot;</span></code> determines the type of covariance matrix of the Gaussian mixture model. It can be <code class="docutils literal notranslate"><span class="pre">&quot;fixed&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;isotropic&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;diagonal&quot;</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;log_var_params&quot;</span></code> stores the parameters to initialize the parameters of the Gaussian prior for the log-variance of the mixture model. The parameters are the mean and standard deviation of the Gaussian prior.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;alpha&quot;</span></code> stores the alpha of the Dirichlet distribution, which determines the uniformity of the weights of the components in the Gaussian mixture.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;mean_prior&quot;</span></code> stores a set of options referring to the prior used for the means of the components in the Gaussian mixture:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;type&quot;</span></code> is the type of prior that should be used. So far, only the <code class="docutils literal notranslate"><span class="pre">&quot;softball&quot;</span></code> prior has been implemented.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;options&quot;</span></code> is a dictionary storing the options that should be used when initializing the prior. The options can vary depending on the <code class="docutils literal notranslate"><span class="pre">&quot;type&quot;</span></code> of prior. For the <code class="docutils literal notranslate"><span class="pre">&quot;softball&quot;</span></code> prior, we have two options:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;radius&quot;</span></code> is the radius of the multi-dimensional soft ball.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;sharpness&quot;</span></code> is the sharpness of the soft boundary of the ball.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;dec&quot;</span></code> maps to a sub-dictionary containing decoder-specific options.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;pth_file&quot;</span></code> indicates the path to the PyTorch file containing the pre-trained decoder. This file is not available inside the GitHub repository because of size constraints, but can be downloaded using <a class="reference external" href="https://drive.google.com/file/d/1SZaoazkvqZ6DBF-adMQ3KRcy4Itxsz77/view?usp=sharing">this link</a>. Once downloaded, you can move this file into the <code class="docutils literal notranslate"><span class="pre">bulkDGD/data/model</span></code> directory and simply refer to it as <code class="docutils literal notranslate"><span class="pre">&quot;dec.pth&quot;</span></code> in the configuration file, instead of using its full path. If only a file name is provided, the <code class="docutils literal notranslate"><span class="pre">get_config_model</span></code> function will automatically look for a file with the specified name in <code class="docutils literal notranslate"><span class="pre">bulkDGD/data/model</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;options&quot;</span></code> maps to a sub-dictionary containing options consistent with those used when training the decoder. Therefore, you should leave them unchanged unless you re-train the decoder using a different set of options.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_neurons_input&quot;</span></code> defines the number of neurons in the input layer of the decoder. This number must correspond to the one set in the <code class="docutils literal notranslate"><span class="pre">&quot;dim&quot;</span></code> option of the <code class="docutils literal notranslate"><span class="pre">[&quot;gmm&quot;][&quot;options&quot;]</span></code> section of the configuration file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_neurons_hidden1&quot;</span></code> defines the number of neurons in the first hidden layer of the decoder.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_neurons_hidden2</span></code> defines the number of neurons in the second hidden layer of the decoder.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_neurons_output&quot;</span></code> defines the number of neurons in the output layer of the decoder.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;r_init&quot;</span></code> defines the initial r-value of each of the negative binomial distributions modeling the expression of the genes.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</section>
<section id="the-configuration-for-finding-the-best-representations">
<h4>The configuration for finding the best representations<a class="headerlink" href="#the-configuration-for-finding-the-best-representations" title="Permalink to this heading">#</a></h4>
<p>We can load the configuration with the options to fine-tune the search for the best representations using the <code class="docutils literal notranslate"><span class="pre">utils.misc.get_config_rep</span></code> function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the configuration with the options to configure</span>
<span class="c1"># the search for the best representations</span>
<span class="n">config_rep</span> <span class="o">=</span> <span class="n">misc</span><span class="o">.</span><span class="n">get_config_rep</span><span class="p">(</span><span class="s2">&quot;config_rep.yaml&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>This configuration also consists of a dictionary of options.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># We can inspect this configuration, too</span>
<span class="n">config_rep</span>
<span class="o">&gt;&gt;&gt;</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;batch_size&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">,</span> <span class="s2">&quot;shuffle&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span> <span class="s2">&quot;rep&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;n_rep_per_comp&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;opt1&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;epochs&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;adam&quot;</span><span class="p">,</span> <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;lr&quot;</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span> <span class="s2">&quot;weight_decay&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;betas&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]}},</span> <span class="s2">&quot;opt2&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;epochs&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;adam&quot;</span><span class="p">,</span> <span class="s2">&quot;options&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;lr&quot;</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span> <span class="s2">&quot;weight_decay&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;betas&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]}}}}</span>
</pre></div>
</div>
<p>Here, we provide a brief description of every option. You can find similar descriptions in the comments inside the configuration file.</p>
<p><code class="docutils literal notranslate"><span class="pre">&quot;data&quot;</span></code> maps to a dictionary containing the options to specify how the samples should be loaded (they are passed to the <a class="reference external" href="https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader"><code class="docutils literal notranslate"><span class="pre">torch.data.utils.DataLoader</span></code></a> constructor):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;batch_size&quot;</span></code> determines the size of the samples’ batches. In fact, samples are loaded in chunks of <code class="docutils literal notranslate"><span class="pre">&quot;batch_size&quot;</span></code> size, and each batch is processed independently.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;shuffle&quot;</span></code> determines whether the samples should be randomly shuffled at every epoch.</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">&quot;rep&quot;</span></code> maps to a dictionary containing the options defining the parameters to search for the best representations for the given samples:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;n_rep_per_comp&quot;</span></code> defines the number of representations to be initialized per component of the Gaussian mixture model per sample.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;opt1&quot;</span></code> maps to a dictionary with the options for the first optimization:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;epochs&quot;</span></code> defines the number of epochs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;type&quot;</span></code> defines the type of optimizer to be used. So far, only the <code class="docutils literal notranslate"><span class="pre">&quot;adam&quot;</span></code> optimizer is supported.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;options&quot;</span></code> maps to a dictionary containing the options to set up the optimizer. These options may vary according to the optimizer used. For <code class="docutils literal notranslate"><span class="pre">&quot;adam&quot;</span></code>, we have three options, which are passed to the <a class="reference external" href="https://pytorch.org/docs/stable/generated/torch.optim.Adam.html"><code class="docutils literal notranslate"><span class="pre">torch.optim.Adam</span></code></a> constructor:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;lr&quot;</span></code> defines the learning rate.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;weight_decay&quot;</span></code> defines the weight decay.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;betas&quot;</span></code> defines the coefficients used for computing running averages of gradient and its square.</p></li>
</ul>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;opt2&quot;</span></code> maps to a dictionary with the options for the second optimization:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;epochs&quot;</span></code> defines the number of epochs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;type&quot;</span></code> defines the type of optimizer to be used. So far, only the <code class="docutils literal notranslate"><span class="pre">&quot;adam&quot;</span></code> optimizer is supported.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;options&quot;</span></code> maps to a dictionary containing the options to set up the optimizer. These options may vary according to the optimizer used. For <code class="docutils literal notranslate"><span class="pre">&quot;adam&quot;</span></code>, we have three options, which are passed to the <a class="reference external" href="https://pytorch.org/docs/stable/generated/torch.optim.Adam.html"><code class="docutils literal notranslate"><span class="pre">torch.optim.Adam</span></code></a> constructor:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;lr&quot;</span></code> defines the learning rate.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;weight_decay&quot;</span></code> defines the weight decay.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;betas&quot;</span></code> defines the coefficients used for computing running averages of gradient and its square.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</section>
</section>
<section id="the-trained-dgd-model">
<h3>The trained DGD model<a class="headerlink" href="#the-trained-dgd-model" title="Permalink to this heading">#</a></h3>
<p>Then, we need to get the trained DGD model. We do so with the <code class="docutils literal notranslate"><span class="pre">get_model</span></code> function in the <code class="docutils literal notranslate"><span class="pre">bulkDGD.utils.dgd</span></code> module.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the trained DGD model (Gaussian mixture model</span>
<span class="c1"># and decoder)</span>
<span class="n">gmm</span><span class="p">,</span> <span class="n">dec</span> <span class="o">=</span> <span class="n">dgd</span><span class="o">.</span><span class="n">get_model</span><span class="p">(</span><span class="n">config_gmm</span> <span class="o">=</span> <span class="n">config_model</span><span class="p">[</span><span class="s2">&quot;gmm&quot;</span><span class="p">],</span>
                         <span class="n">config_dec</span> <span class="o">=</span> <span class="n">config_model</span><span class="p">[</span><span class="s2">&quot;dec&quot;</span><span class="p">])</span>

<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">latent</span><span class="p">:</span><span class="n">The</span> <span class="n">SoftballPrior</span> <span class="n">prior</span> <span class="n">will</span> <span class="n">be</span> <span class="n">used</span> <span class="k">as</span> <span class="n">prior</span> <span class="n">over</span> <span class="n">the</span> <span class="n">means</span> <span class="n">of</span> <span class="n">the</span> <span class="n">mixture</span> <span class="n">components</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">The</span> <span class="n">Gaussian</span> <span class="n">mixture</span> <span class="n">model</span> <span class="n">was</span> <span class="n">successfully</span> <span class="n">initialized</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">The</span> <span class="n">Gaussian</span> <span class="n">mixture</span> <span class="n">model</span><span class="s1">&#39;s state was successfully loaded from &#39;</span><span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">testuser</span><span class="o">/</span><span class="n">programs</span><span class="o">/</span><span class="n">bulkDGD</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">model</span><span class="o">/</span><span class="n">gmm</span><span class="o">.</span><span class="n">pth</span><span class="s1">&#39;.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">The</span> <span class="n">decoder</span> <span class="n">was</span> <span class="n">successfully</span> <span class="n">initialized</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">The</span> <span class="n">decoder</span><span class="s1">&#39;s state was successfully loaded from &#39;</span><span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">testuser</span><span class="o">/</span><span class="n">programs</span><span class="o">/</span><span class="n">bulkDGD</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">model</span><span class="o">/</span><span class="n">dec</span><span class="o">.</span><span class="n">pth</span><span class="s1">&#39;.</span>
</pre></div>
</div>
<p>As we can see from the log messages, both the Gaussian mixture model and the decoder were first initialized, and then their state (the trained parameters) were loaded.</p>
</section>
<section id="the-representations">
<h3>The representations<a class="headerlink" href="#the-representations" title="Permalink to this heading">#</a></h3>
<p>We can now use the <code class="docutils literal notranslate"><span class="pre">get_representations</span></code> function in the <code class="docutils literal notranslate"><span class="pre">bulkDGD.utils.dgd</span></code> module to find the best representations for our samples of interest.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the representations</span>
<span class="n">df_rep</span><span class="p">,</span> <span class="n">df_dec_out</span> <span class="o">=</span> \
    <span class="n">dgd</span><span class="o">.</span><span class="n">get_representations</span><span class="p">(</span>\
        <span class="c1"># The data frame containing the preprocessed samples</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df_preproc</span><span class="p">,</span>
        <span class="c1"># The trained Gaussian mixture model</span>
        <span class="n">gmm</span> <span class="o">=</span> <span class="n">gmm</span><span class="p">,</span>
        <span class="c1"># The trained decoder</span>
        <span class="n">dec</span> <span class="o">=</span> <span class="n">dec</span><span class="p">,</span>
        <span class="c1"># How many representations to initialize per component</span>
        <span class="c1"># of the Gaussian mixture model per sample</span>
        <span class="n">n_rep_per_comp</span> <span class="o">=</span> <span class="n">config_rep</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">][</span><span class="s2">&quot;n_rep_per_comp&quot;</span><span class="p">],</span>
        <span class="c1"># The configuration to load the samples</span>
        <span class="n">config_data</span> <span class="o">=</span> <span class="n">config_rep</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">],</span>
        <span class="c1"># The configuration for the first optimization</span>
        <span class="n">config_opt1</span> <span class="o">=</span> <span class="n">config_rep</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">][</span><span class="s2">&quot;opt1&quot;</span><span class="p">],</span>
        <span class="c1"># The configuration for the second optimization</span>
        <span class="n">config_opt2</span> <span class="o">=</span> <span class="n">config_rep</span><span class="p">[</span><span class="s2">&quot;rep&quot;</span><span class="p">][</span><span class="s2">&quot;opt2&quot;</span><span class="p">])</span>

<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Starting</span> <span class="n">the</span> <span class="n">first</span> <span class="n">optimization</span><span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">1</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">53.51452738405719</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">2</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">37.78411296090515</span><span class="o">.</span>
<span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">10</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">21.375471316397174</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Starting</span> <span class="n">the</span> <span class="n">second</span> <span class="n">optimization</span><span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">1</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">16.999307299157767</span><span class="o">.</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">2</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">16.984738363634538</span><span class="o">.</span>
<span class="o">...</span>
<span class="n">INFO</span><span class="p">:</span><span class="n">bulkDGD</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dgd</span><span class="p">:</span><span class="n">Epoch</span> <span class="mi">50</span><span class="p">:</span> <span class="n">loss</span> <span class="mf">16.913638140095696</span><span class="o">.</span>
</pre></div>
</div>
<p>We can see from the log messages how the optimization is proceeding and inspect the per-epoch total loss.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">get_representations</span></code> function returns two objects:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">df_rep</span></code> is a data frame containing the representations. It has the following structure:</p>
<p>Each row contains the best representation found for each sample, with the first column containing the name/ID/index of the sample as provided in the input data frame. The other columns contain the values of the samples’ representation along the latent space’s dimensions, the loss associated with each representation (<code class="docutils literal notranslate"><span class="pre">loss</span></code> column), and the additional information found for the samples in the input data frame.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">df_dec_out</span></code> is a data frame containing the outputs of the decoder for the representations found.</p>
<p>As for <code class="docutils literal notranslate"><span class="pre">df_rep</span></code>, each row represents the best representation found for each sample, with the first column containing the name/ID/index of the sample as provided in the input data frame. The other columns here contain the means of the negative binomials modeling the expression of each gene in each sample according to the DGD model and the additional information found for the samples in the input data frame.</p>
</li>
</ul>
<p>The r-values of the negative binomials can be obtained with the <code class="docutils literal notranslate"><span class="pre">bulkDGD.utils.dgd.get_r_values</span></code> function, which gets them from the trained decoder.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the r-values</span>
<span class="n">r_values</span> <span class="o">=</span> <span class="n">dgd</span><span class="o">.</span><span class="n">get_r_values</span><span class="p">(</span><span class="n">dec</span> <span class="o">=</span> <span class="n">dec</span><span class="p">)</span>
</pre></div>
</div>
<p>We can now save the representations and decoder outputs into CSV files using the <code class="docutils literal notranslate"><span class="pre">save_representations</span></code> and <code class="docutils literal notranslate"><span class="pre">save_decoder_outputs</span></code> in the <code class="docutils literal notranslate"><span class="pre">bulkDGD.utils.dgd</span></code> module.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save the representations</span>
<span class="n">dgd</span><span class="o">.</span><span class="n">save_representations</span><span class="p">(</span>\
    <span class="c1"># The data frame containing the representations</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df_rep</span><span class="p">,</span>
    <span class="c1"># The output CSV file</span>
    <span class="n">csv_file</span> <span class="o">=</span> <span class="s2">&quot;representations.csv&quot;</span><span class="p">,</span>
    <span class="c1"># The field separator in the output CSV file</span>
    <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span>

<span class="c1"># Save the decoder outputs</span>
<span class="n">dgd</span><span class="o">.</span><span class="n">save_decoder_outputs</span><span class="p">(</span>\
    <span class="c1"># The data frame containing the decoder outputs</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df_dec_out</span><span class="p">,</span>
    <span class="c1"># The output CSV file</span>
    <span class="n">csv_file</span> <span class="o">=</span> <span class="s2">&quot;decoder_outputs.csv&quot;</span><span class="p">,</span>
    <span class="c1"># The field separator in the output CSV file</span>
    <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="tutorial-2-differential-expression-analysis">
<h2>Tutorial 2 - Differential expression analysis<a class="headerlink" href="#tutorial-2-differential-expression-analysis" title="Permalink to this heading">#</a></h2>
<p>[coming soon]</p>
</section>
<section id="tutorial-3-principal-component-analysis">
<h2>Tutorial 3 - Principal component analysis<a class="headerlink" href="#tutorial-3-principal-component-analysis" title="Permalink to this heading">#</a></h2>
<p>[coming soon]</p>
</section>
</section>


                </article>
              
              
              
                <footer class="bd-footer-article">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="Command-line%20interface.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">The command-line interface</p>
      </div>
    </a>
    <a class="right-next"
       href="core.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">The core subpackage</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial-1-finding-the-best-representations-for-a-set-of-new-samples">Tutorial 1 - Finding the best representations for a set of new samples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-input-samples">The input samples</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-configuration-files">The configuration files</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-model-s-configuration">The model’s configuration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-configuration-for-finding-the-best-representations">The configuration for finding the best representations</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-trained-dgd-model">The trained DGD model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-representations">The representations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial-2-differential-expression-analysis">Tutorial 2 - Differential expression analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial-3-principal-component-analysis">Tutorial 3 - Principal component analysis</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div class="tocsection sourcelink">
    <a href="_sources/Tutorials.md.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            <div class="bd-footer-content__inner"></div>
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2023, Valentina Sora.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://sphinx-doc.org/">Sphinx</a> 6.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.13.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>